python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -414,51 +426,137 @@ def execute_asset_backfill_iteration(
-        asset_backfill_data=asset_backfill_data,
-        instance=instance,
-        asset_graph=asset_graph,
-        run_tags=backfill.tags,
-        backfill_start_time=utc_datetime_from_timestamp(backfill.backfill_timestamp),
-    ):
-        yield None
+    instance_queryer = CachingInstanceQueryer(
+        instance=instance, evaluation_time=backfill_start_time
+    )
 
-    if not isinstance(result, AssetBackfillIterationResult):
-        check.failed(
-            "Expected execute_asset_backfill_iteration_inner to return an"
-            " AssetBackfillIterationResult"
+    if backfill.status == BulkActionStatus.REQUESTED:
+        result = None
+        for result in execute_asset_backfill_iteration_inner(
+            backfill_id=backfill.backfill_id,
+            asset_backfill_data=asset_backfill_data,
+            instance_queryer=instance_queryer,
+            asset_graph=asset_graph,
+            run_tags=backfill.tags,
+            backfill_start_time=backfill_start_time,
+        ):
+            yield None
+
+        if not isinstance(result, AssetBackfillIterationResult):
+            check.failed(
+                "Expected execute_asset_backfill_iteration_inner to return an"
+                " AssetBackfillIterationResult"
+            )
+
+        updated_backfill = backfill.with_asset_backfill_data(
+            result.backfill_data, dynamic_partitions_store=instance
         )
+        if result.backfill_data.is_complete():
+            # The asset backfill is complete when all runs to be requested have finished (success,
+            # failure, or cancellation). Since the AssetBackfillData object stores materialization states
+            # per asset partition, the daemon continues to update the backfill data until all runs have
+            # finished in order to display the final partition statuses in the UI.
+            updated_backfill = updated_backfill.with_status(BulkActionStatus.COMPLETED)
 
-    updated_backfill = backfill.with_asset_backfill_data(
-        result.backfill_data, dynamic_partitions_store=instance
-    )
-    if result.backfill_data.is_complete():
-        # The asset backfill is complete when all runs to be requested have finished (success,
+        pipeline_and_execution_plan_cache: Dict[int, Tuple[ExternalJob, ExternalExecutionPlan]] = {}
+
+        for run_request in result.run_requests:
+            yield None
+            submit_run_request(
+                run_request=run_request,
+                asset_graph=asset_graph,
+                # create a new request context for each run in case the code location server
+                # is swapped out in the middle of the backfill
+                workspace=workspace_process_context.create_request_context(),
+                instance=instance,
+                pipeline_and_execution_plan_cache=pipeline_and_execution_plan_cache,
+            )
+
+    elif backfill.status == BulkActionStatus.CANCELING:
```

Comment:
Thoughts on pushing this logic down into `execute_asset_backfill_iteration_inner`?

The way I conceived the difference between `execute_asset_backfill_iteration_inner` and `execute_asset_backfill_iteration` is that `execute_asset_backfill_iteration_inner` includes all the "smarts", without doing any side effects, and `execute_asset_backfill_iteration` carries out the side effects, i.e. submitting runs, updating the backfill record in the database.

Maintaining this separation makes it possible to test the smarts without dealing with the side effects.

It seems like the cancellation logic doesn't have any side effects, so could go in the inner function?

Reply:
There's been some changes to this function since you last reviewed, I think now there are side effects--runs are canceled, and the bulk action status is updated.

Let me know if you'd still prefer a refactor here